keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,171 @@
+
+Kubernetes cgroup manager uses `memory.limit_in_bytes` in cgroups v1 and `memory.max` in cgroups v2 to limit memory capacity and uses `oom_scores` to determine order of killing container process when OOM occurs.
+
+Above implement has following disadvantages:
+- `Guaranteed` pod memory can not be fully reserved, page cache is at risk of being recycled
+- `Burstable` pod memory allocation is not well protected, there may occur allocation latency frequently when node memory nearly runs out
+
+Currently we only use `memory.limit_in_bytes = sum(pod.spec.containers.resources.limits[memory])` to limit memory hardly in cgroups v1 and `memory.max = sum(pod.spec.containers.resources.limits[memory])` in cgroups v2. And `resources.requests[memory]` is not yet used neither cgroups v1 nor cgroups v2.
+
+cgroups v2 introduces a better way to limit and guarantee memory quality.
+
+| File | Description |
+| -------- | -------- |
+| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup’s memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. |
+| memory.max | memory.max is the memory usage hard limit, acting as the final protection mechanism: If a cgroup's memory usage reaches this limit and can't be reduced, the system OOM killer is invoked on the cgroup. Under certain circumstances, usage may go over the memory.high limit temporarily. When the high limit is used and monitored properly, memory.max serves mainly to provide the final safety net. The default is max. |
+| memory.low | memory.low is the best-effort memory protection, a “soft guarantee” that if the cgroup and all its descendants are below this threshold, the cgroup's memory won't be reclaimed unless memory can’t be reclaimed from any unprotected cgroups. |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup’s memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. |
+
+This proposal uses `memory.min` to retain request memory for `Guaranteed` pod. And for `Burstable` pod, it uses `memory.low` to achieve best-effort memory protection by summing container `requests.memory` in pod. We also set `memory.max` to do hard limit to make OOM killer work.
+
+### User Stories (Optional)
+- Workload is memory-sensitive, this feature can retain memory to reduce allocation latency
+- Workload is burstable, this feature can try best to protect request memory and decrease probability of memory being reclaimed
+
+### Risks and Mitigations
+- `Guaranteed` pod memory reservation could reduce node memory supply
+
+## Design Details
+### Feature Gate
+To set `--feature-gates=MemoryQOS=true` to enable the feature.
+
+### CGroup Resource
```

Comment:
For the KEP, I think you can omit the specific code changes, its hard to reason about until we get to implementation.

I really like the proposed mapping though.

Is there anything we can do at the node allocatable cgroup level with memory that you would recommend?

Reply:
I think what we do at that level will depend on what we want to protect more. One possible option is to use `memory.low` for system reserved. That way the guaranteed and burstable will be protected by the proposed memory.min and memory.low and the best effort would be unprotected and first targets for memory reclamation.

Reply:
> For the KEP, I think you can omit the specific code changes, its hard to reason about until we get to implementation.
ok, I will remove code details.

Reply:
> I think what we do at that level will depend on what we want to protect more. One possible option is to use `memory.low` for system reserved. That way the guaranteed and burstable will be protected by the proposed `memory.min` and `memory.low` and the best effort would be unprotected and first targets for memory reclamation.

agree, we can set `memory.low` for system reserved and kube reserved.

Reply:
> For the KEP, I think you can omit the specific code changes, its hard to reason about until we get to implementation.
> 
> I really like the proposed mapping though.
> 
> Is there anything we can do at the node allocatable cgroup level with memory that you would recommend?

Updated.